src/main/java/appeng/core/features/registries/P2PTunnelRegistry.java

```diff
@@ -203,39 +223,40 @@ public TunnelType getTunnelTypeByItem( final ItemStack trigger )
 	{
 		if( !trigger.isEmpty() )
 		{
-			// if( FluidRegistry.isContainer( trigger ) )
-			// {
-			// return TunnelType.FLUID;
-			// }
-
-			for( final ItemStack is : this.tunnels.keySet() )
+			// First match exact items
+			for( final Entry<ItemStack, TunnelType> entry : this.tunnels.entrySet() )
 			{
+				final ItemStack is = entry.getKey();
+
 				if( is.getItem() == trigger.getItem() && is.getItemDamage() == OreDictionary.WILDCARD_VALUE )
 				{
-					return this.tunnels.get( is );
+					return entry.getValue();
 				}
 
 				if( ItemStack.areItemsEqual( is, trigger ) )
 				{
-					return this.tunnels.get( is );
+					return entry.getValue();
 				}
 			}
 
-			// Try by ModId next
-			for( final String modId : this.modIdTunnels.keySet() )
+			// Next, check if the Item you're holding supports any registered capability
+			for( EnumFacing face : EnumFacing.VALUES )
 			{
-				if( trigger.getItem().getRegistryName() != null && trigger.getItem().getRegistryName().getResourceDomain().equals( modId ) )
+				for( Entry<Capability<?>, TunnelType> entry : this.capTunnels.entrySet() )
 				{
-					return this.modIdTunnels.get( modId );
+					if( trigger.hasCapability( entry.getKey(), face ) )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)